pkg/runner/hash_map.go

```diff
@@ -0,0 +1,96 @@
+package runner
+
+import "sync"
+
+type hashMap struct {
+	mut    sync.RWMutex
+	lookup map[uint64][]Task
+}
+
+// newHashMap creates a new hashMap allocated to handle at least size unique
+// hashes.
+func newHashMap(size int) *hashMap {
+	return &hashMap{
+		lookup: make(map[uint64][]Task, size),
+	}
+}
+
+// Add adds the provided Task into the hashMap. It returns false if t
+// already exists in the hashMap.
+func (hm *hashMap) Add(t Task) bool {
+	hm.mut.Lock()
+	defer hm.mut.Unlock()
+
+	hash := t.Hash()
+	for _, compare := range hm.lookup[hash] {
+		if compare.Equals(t) {
+			return false
+		}
+	}
+
+	hm.lookup[hash] = append(hm.lookup[hash], t)
+	return true
+}
+
+// Has returns true if t exists in the hashMap.
+func (hm *hashMap) Has(t Task) bool {
+	hm.mut.RLock()
+	defer hm.mut.RUnlock()
+
+	for _, compare := range hm.lookup[t.Hash()] {
+		if compare.Equals(t) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Delete removes the provided Task from the hashMap. It returns true if the
+// Task was found and deleted.
+func (hm *hashMap) Delete(t Task) (deleted bool) {
+	hm.mut.Lock()
+	defer hm.mut.Unlock()
+
+	hash := t.Hash()
+
+	var rem []Task
+	for _, s := range hm.lookup[hash] {
+		if s.Equals(t) {
+			deleted = true
```

Comment:
This feels off in the naming. I read `rem` as `remove`, but I think it means `remaining`?

Reply:
Yep, I'll rename 👍